Add offers table and model

rshbhgrg opened the pull request:
refs https://github.com/TryGhost/Team/issues/1088

- adds new offer table and model
- adds migrations

allouis commented on core/server/data/migrations/versions/4.17/02-add-offers-table.js:
```diff
@@ -0,0 +1,17 @@
+    slug: {type: 'string', maxlength: 191, nullable: false, unique: true},
+    product_id: {type: 'string', maxlength: 24, nullable: false},
+    interval: {type: 'string', maxlength: 50, nullable: false, validations: {isIn: [['monthly', 'yearly']]}},
+    currency: {type: 'string', maxlength: 50, nullable: false},
```
Percentage based offers won't have a currency - so I think we need to either have this as `nullable` or we store like `'ANY'` or something for the currency - what do you think?

rishabhgrg replied:
Yea i had the thought, and my feeling was we can always just store the currency for the price at the time offer was created irrespective if it was for percentage based or amount based. For percentage based it will simply get ignored in the logic, while for amount based we can use it, but overall the logic can remain the same and we don't need any extra work for either type, wdyt ?

allouis replied:
Sure we can store it if you think we should ☺️

rishabhgrg replied:
Have updated the `nullable` to `true` in this case as on second thought it does make sense to leave the option in DB to not store anything for it as you said, we can always ensure in logic we save currency for both anyways :)

allouis commented on core/server/data/migrations/versions/4.17/02-add-offers-table.js:
```diff
@@ -0,0 +1,17 @@
+    currency: {type: 'string', maxlength: 50, nullable: false},
+    discount_type: {type: 'string', maxlength: 50, nullable: false},
+    discount_amount: {type: 'integer', nullable: false},
+    duration: {type: 'string', maxlength: 50, nullable: false},
```
If we have a duration of `'repeating'` then we need to store for how long it repeats (e.g. 3 months) so I think we either need another column of like `duration_count` or something or we store a number in this column 🤔 

`0` - once
`Infinity` - forever
`1`,`2`,...`1000` - number of months

rishabhgrg replied:
🤔 Good catch, i missed the need for storing `duration_count` in case of `repeating`, i'll add a nullable `duration_in_months ` integer column that can store the duration count in number of months to stay consistent with stripe, don't think overloading the `duration` column with number is good

rishabhgrg replied:
Added `duration_in_months` column now

allouis commented on core/server/data/migrations/versions/4.17/02-add-offers-table.js:
```diff
@@ -0,0 +1,17 @@
+module.exports = utils.addTable('offers', {
+    id: {type: 'string', maxlength: 24, nullable: false, primary: true},
+    name: {type: 'string', maxlength: 191, nullable: false},
+    slug: {type: 'string', maxlength: 191, nullable: false, unique: true},
```
Is this to store the Code for an Offer?

rishabhgrg replied:
Yeah this is to store the offer code, i realized the offer code is essentially as good as an offer slug since it will have all of the slug properties for a URL, so we can just store it as that. Was split between naming it `slug` or something like `code`, but `slug` seemed more consistent to me so went with that, not a strong pref tho, wdyt ?

allouis replied:
It wasn't immediately obvious to me what this was storing - but happy to leave it like this if it makes sense to you

rishabhgrg replied:
It not being immediately obvious is definitely a problem, maybe it makes more sense to use the name as `code` in this case so we don't have same problem down the link. I'll rename it to `code` so its more clear, we can always make it behave like a slug behind the scenes :)

rishabhgrg replied:
Done

allouis commented on core/server/models/offer.js:
```diff
@@ -0,0 +1,53 @@
+            model.set('name', model.get('name').trim());
+        }
+
+        if (model.hasChanged('slug') || !model.get('slug')) {
```
I'm wondering if we need this logic too - I think this can be handled by Offers module if necessary too!

rishabhgrg replied:
🤔 Yeah this is basically a copy of old practice we have been following so far for slugs, we could move it to offers module tho which might be cleaner. We'll essentially just need to duplicate the logic in `generateSlug` plugin method, but might be a cleaner cut so on same page

rishabhgrg replied:
Trimmed down models now so we can move this logic to members codebase

allouis replied:
Cool - sounds good!

allouis commented on core/server/models/offer.js:
```diff
@@ -0,0 +1,53 @@
+    }
+});
+
+const Offers = ghostBookshelf.Collection.extend({
```
Can remove this line - we're not adding collection going forward!

rishabhgrg replied:
Ah ok cool, updating 👍

rishabhgrg replied:
Done

allouis commented on core/server/models/offer.js:
```diff
@@ -0,0 +1,53 @@
+    async onSaving(model, _attr, options) {
+        ghostBookshelf.Model.prototype.onSaving.apply(this, arguments);
+
+        if (model.get('name')) {
```
I think this can be handled by the Offers module so all the logic is in one place - wdyt?

rishabhgrg replied:
Comment below ⬇️

rishabhgrg replied:
Done

allouis reviewed (commented):
Looks good - few comments and questions!

allouis commented on core/server/data/exporter/table-lists.js:
```diff
@@ -25,6 +25,7 @@ const BACKUP_TABLES = [
     'snippets',
     'tokens',
     'sessions',
+    'offers',
```
I always forget this!

allouis commented on core/server/data/migrations/versions/4.17/03-add-offers-permissions.js:
```diff
@@ -0,0 +1,42 @@
+    }, [
+        'Administrator'
+    ]),
+    addPermissionWithRoles({
```
Do we need/want this permission? AFAIK We're only going to implement archive not delete

rishabhgrg replied:
🤔 Hmm yeah thats true, can bring it back if/when needed, will remove it for now

rishabhgrg replied:
Removed

allouis commented on core/server/data/migrations/versions/4.17/02-add-offers-table.js:
```diff
@@ -0,0 +1,18 @@
+    name: {type: 'string', maxlength: 191, nullable: false},
+    code: {type: 'string', maxlength: 191, nullable: false, unique: true},
+    product_id: {type: 'string', maxlength: 24, nullable: false},
+    interval: {type: 'string', maxlength: 50, nullable: false, validations: {isIn: [['monthly', 'yearly']]}},
```
These intervals are incorrect btw - it should be `'month'` and `'year'`

rishabhgrg replied:
👍 Updated

allouis commented on core/server/data/migrations/versions/4.17/02-add-offers-table.js:
```diff
@@ -0,0 +1,18 @@
+const utils = require('../../utils');
+
+module.exports = utils.addTable('offers', {
```
Oh I just realised - this is missing a column for `stripe_coupon_id`!

rishabhgrg replied:
Done

rshbhgrg committed:
Added `offers` table

refs https://github.com/TryGhost/Team/issues/1088

- adds schema for new offers table
- adds permission fixtures for new offers table
- adds migrations for new table and permissions

rshbhgrg committed:
Updated tests for `offers` table

refs https://github.com/TryGhost/Team/issues/1088

- updates tests for new `offers` table and permissions

rshbhgrg committed:
Added model for `offers` table

refs https://github.com/TryGhost/Team/issues/1088

- adds basic model definition for `offers` table

rishabhgrg commented:
Closed in favor of https://github.com/TryGhost/Ghost/pull/13381

allouis commented on core/server/models/offer.js:
```diff
@@ -0,0 +1,13 @@
+const ghostBookshelf = require('./base');
+
+const Offer = ghostBookshelf.Model.extend({
```
This is missing being exported from `models/index.js` btw!
